# ADS 509 Module 1: APIs and Web Scraping
## Sarah Alqaysi

This notebook has three parts. In the first part you will pull data from the Twitter API. In the second, you will scrape lyrics from AZLyrics.com. In the last part, you'll run code that verifies the completeness of your data pull. 

For this assignment you have chosen two musical artists who have at least 100,000 Twitter followers and 20 songs with lyrics on AZLyrics.com. In this part of the assignment we pull the some of the user information for the followers of your artist and store them in text files. 


## General Assignment Instructions

These instructions are included in every assignment, to remind you of the coding standards for the class. Feel free to delete this cell after reading it. 

One sign of mature code is conforming to a style guide. We recommend the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html). If you use a different style guide, please include a cell with a link. 

Your code should be relatively easy-to-read, sensibly commented, and clean. Writing code is a messy process, so please be sure to edit your final submission. Remove any cells that are not needed or parts of cells that contain unnecessary code. Remove inessential `import` statements and make sure that all such statements are moved into the designated cell. 

Make use of non-code cells for written commentary. These cells should be grammatical and clearly written. In some of these cells you will have questions to answer. The questions will be marked by a "Q:" and will have a corresponding "A:" spot for you. *Make sure to answer every question marked with a `Q:` for full credit.* 


# Twitter API Pull

In [1]:
# for the twitter section
import tweepy
import os
import datetime
import re
from pprint import pprint

# for the lyrics scrape section
import requests
import time
from bs4 import BeautifulSoup
from collections import defaultdict, Counter


In [2]:
# Use this cell for any import statements you add

import warnings

# Suppress warning messages
warnings.filterwarnings('ignore')

#!pip install git+https://github.com/tweepy/tweepy.git

We need bring in our API keys. Since API keys should be kept secret, we'll keep them in a file called `api_keys.py`. This file should be stored in the directory where you store this notebook. The example file is provided for you on Blackboard. The example has API keys that are _not_ functional, so you'll need to get Twitter credentials and replace the placeholder keys. 

In [3]:
from api_keys import api_key, api_key_secret, bearer_token 

In [12]:
auth = tweepy.AppAuthHandler(api_key, api_key_secret)
api = tweepy.API(auth,wait_on_rate_limit=True)
                 #wait_on_rate_limit_notify=True)

print ('API Host', api.host)
#print ('API Version', api.api_root)

API Host api.twitter.com


## Testing the API

The Twitter APIs are quite rich. Let's play around with some of the features before we dive into this section of the assignment. For our testing, it's convenient to have a small data set to play with. We will seed the code with the handle of John Chandler, one of the instructors in this course. His handle is `@37chandler`. Feel free to use a different handle if you would like to look at someone else's data. 

We will write code to explore a few aspects of the API: 

1. Pull all the follower IDs for @katymck.
1. Explore the user object, which gives us information about Twitter users. 
1. Pull some user objects for the followers. 
1. Pull the last few tweets by @katymck.


In [4]:
client = tweepy.Client(bearer_token,wait_on_rate_limit=True)

In [95]:
handle = "37chandler"
user_obj = client.get_user(username=handle)

followers = client.get_users_followers(
    # Learn about user fields here: 
    # https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/user
    user_obj.data.id, user_fields=["created_at","description","location",
                                   "public_metrics"]
)

We have the follower IDs, which are unique numbers identifying the user, but we'd like to get some more information on these users. Twitter allows us to pull "fully hydrated user objects", which is a fancy way of saying "all the information about the user". Let's look at user object for our starting handle.

In [99]:
api = tweepy.API(auth)
user = api.get_user(screen_name='handle')

print(user._json)

{'id': 382247773, 'id_str': '382247773', 'name': 'Handle', 'screen_name': 'Handle', 'location': 'San Francisco, CA', 'profile_location': None, 'description': 'We believe in getting the most out of our time, and saw that one of the most-used tools on the planet, the email inbox, was failing us. Re-Branding!', 'url': 'https://t.co/sQdLIf7LR3', 'entities': {'url': {'urls': [{'url': 'https://t.co/sQdLIf7LR3', 'expanded_url': 'http://handle.today', 'display_url': 'handle.today', 'indices': [0, 23]}]}, 'description': {'urls': []}}, 'protected': False, 'followers_count': 0, 'friends_count': 0, 'listed_count': 0, 'created_at': 'Thu Sep 29 18:53:39 +0000 2011', 'favourites_count': 0, 'utc_offset': None, 'time_zone': None, 'geo_enabled': False, 'verified': False, 'statuses_count': 10, 'lang': None, 'status': {'created_at': 'Mon Apr 08 12:05:42 +0000 2019', 'id': 1115224193715466240, 'id_str': '1115224193715466240', 'text': '@brucedaisley It’s not published yet, will be April 10th! We could link 

Let's pull some more user fields and take a look at them. The fields can be specified in the `user_fields` argument. 

In [100]:
response = client.get_user(id=user_obj.data.id,
                          user_fields=["created_at","description","location",
                                       "entities","name","pinned_tweet_id","profile_image_url",
                                       "verified","public_metrics"])

In [101]:
for field, value in response.data.items() :
    print(f"for {field} we have {value}")

for username we have 37chandler
for id we have 33029025
for verified we have False
for description we have He/Him. Data scientist, urban cyclist, educator, erstwhile frisbee player. 

¯\_(ツ)_/¯
for profile_image_url we have https://pbs.twimg.com/profile_images/2680483898/b30ae76f909352dbae5e371fb1c27454_normal.png
for name we have John Chandler
for location we have MN
for created_at we have 2009-04-18 22:08:22+00:00
for public_metrics we have {'followers_count': 189, 'following_count': 574, 'tweet_count': 945, 'listed_count': 3}


Now let's explore these a bit. We'll start by printing out names, locations, following count, and followers count for these users.

In [102]:
num_to_print = 20

for idx, user in enumerate(followers.data) :
    following_count = user.public_metrics['following_count']
    followers_count = user.public_metrics['followers_count']
    
    print(f"{user.name} lists '{user.location}' as their location.")
    print(f" Following: {following_count}, Followers: {followers_count}.")
    print()
    
    if idx >= (num_to_print - 1) :
        break

Steve lists 'I'm over here.' as their location.
 Following: 1548, Followers: 33.

John O'Connor 🇺🇦 lists 'None' as their location.
 Following: 8, Followers: 1.

CodeGrade lists 'Amsterdam' as their location.
 Following: 2712, Followers: 389.

Cleverhood lists 'Providence, RI' as their location.
 Following: 2775, Followers: 3370.

Paul Naish lists 'Oxford' as their location.
 Following: 19365, Followers: 19549.

Regina 🚲🌳 (end car supremacy) lists 'Minneapolis' as their location.
 Following: 2652, Followers: 2790.

Eric Hallstrom lists 'Missoula, MT' as their location.
 Following: 458, Followers: 304.

Tyler 📊 🐕 🚲 lists 'None' as their location.
 Following: 440, Followers: 48.

The Center for Community Ownership (CCO) lists 'None' as their location.
 Following: 47, Followers: 31.

Deepak Chauhan lists 'None' as their location.
 Following: 575, Followers: 31.

Patsy lists 'Seattle, WA' as their location.
 Following: 153, Followers: 15.

andrew lists 'St Paul, MN' as their location.
 Foll

Let's find the person who follows this handle who has the most followers.

In [103]:
max_followers = 0

for idx, user in enumerate(followers.data) :
    followers_count = user.public_metrics['followers_count']
    
    if followers_count > max_followers :
        max_followers = followers_count
        max_follower_user = user

        
print(max_follower_user)
print(max_follower_user.public_metrics)

JeffreyHayzlett
{'followers_count': 533219, 'following_count': 444897, 'tweet_count': 47991, 'listed_count': 5650}


Now a few questions for you about the user object.

Q: How many fields are being returned in the \_json portion of the user object? 

A: Nine 

---

Q: Are any of the fields within the user object non-scaler? TK correct term

A: Yes. 

---

Q: How many friends, followers, favorites, and statuses does this user have? 

A: {'followers_count': 533219, 'following_count': 444897, 'tweet_count': 47991, 'listed_count': 5650}


We can map the follower IDs onto screen names by accessing the screen_name key within the user object. Modify the code below to also print out how many people the follower is following and how many followers they have. 

In [15]:
followers = []
api.lookup_users(screen_name = handle)

[User(_api=<tweepy.api.API object at 0x000001998D7BFE48>, _json={'id': 287253599, 'id_str': '287253599', 'name': 'Blair', 'screen_name': '3', 'location': 'London, England', 'description': 'Is it Friday yet?', 'url': None, 'entities': {'description': {'urls': []}}, 'protected': False, 'followers_count': 154752, 'friends_count': 102, 'listed_count': 3223, 'created_at': 'Sun Apr 24 16:36:42 +0000 2011', 'favourites_count': 91, 'utc_offset': None, 'time_zone': None, 'geo_enabled': False, 'verified': False, 'statuses_count': 2152, 'lang': None, 'status': {'created_at': 'Sat Jan 16 08:40:16 +0000 2021', 'id': 1350362220304031744, 'id_str': '1350362220304031744', 'text': 'My favourite player of all time. Absolute legend. Excited for his new career. Would love to see him back in the Pre… https://t.co/dLXtvsfqXr', 'truncated': True, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [{'url': 'https://t.co/dLXtvsfqXr', 'expanded_url': 'https://twitter.com/i/web/status/13503

In [20]:
# fetching the favourites_count for "37chandler"
favourites_count = user.favourites_count

# fetching the statuses_count for "37chandler"
statuses_count = user.statuses_count
  
print("The number of tweets the user has liked are : " + str(favourites_count))
print("The number of statuses the user has had are : " + str(statuses_count))


The number of tweets the user has liked are : 1786
The number of statuses the user has had are : 8731


In [17]:
# finding the count of how many followers & friends "37chandler" has

# fetching the user
user = api.get_user(screen_name='37chandler')
  
# fetching the followers_count
followers_count = user.followers_count

# fetching the friends_count
friends_count = user.friends_count

  
print("The number of followers of 37chandler are : " + str(followers_count))
print("The number of friends of 37chandler are : " + str(friends_count))

The number of followers of 37chandler are : 189
The number of friends of 37chandler are : 574


In [18]:
# Add code here to print out friends and followers of `handle`

for user in tweepy.Cursor(api.get_followers, screen_name="37chandler").items():
    print('followers: ' + user.screen_name)

for user in tweepy.Cursor(api.get_friends, screen_name="37chandler").items():
    print('friends: ' + user.screen_name)

followers: HicSvntDraconez
followers: JohnOCo70713197
followers: CodeGradeCom
followers: cleverhoods
followers: PaulNaish78
followers: mplsFietser
followers: echallstrom
followers: byler_t117
followers: Community_Owner
followers: DeepakC64237257
followers: patsy16sb
followers: andrewc186
followers: _ada_smith
followers: WentRogue
followers: MplsBikeTrails
followers: KristinaHubbard
followers: erica_zurek
followers: evelindylan1
followers: yofraze
followers: JenniJenniG
followers: RisaForParks
followers: l7ashley
followers: dodgemc
followers: oldschoolCUT
followers: TMolteni
followers: MeredithWestin
followers: metageniglobal
followers: scoover
followers: QuonkQuonk
followers: rickbruner
followers: peterschmitt34
followers: stalinnaveen
followers: afhauer
followers: mediantweets
followers: gregtenold
followers: MadameDavey
followers: freyjahartz
followers: johnlouisM
followers: FenrisWinston
followers: CogentAg
followers: Worse_Reviewer
followers: puadmsara
followers: davebts77
follower

Although you won't need it for this assignment, individual tweets (called "statuses" in the API) can be a rich source of text-based data. To illustrate the concepts, let's look at the last few tweets for this user. You are encouraged to explore the `status` object and marvel in the richness of the data that is available. 


In [21]:
tweet_count = 0

for status in tweepy.Cursor(api.user_timeline, id=handle).items():
    tweet_count += 1
    
    print(f"The tweet was tweeted at {status.created_at}.")
    print(f"The original tweet has been retweeted {status.retweet_count} times.")
    
    clean_status = status.text
    clean_status = clean_status.replace("\n"," ")
    
    print(f"{clean_status}")
    print("\n"*2)
        
    if tweet_count > 10 :
        break



Unexpected parameter: id


The tweet was tweeted at 2022-05-18 19:43:54+00:00.
The original tweet has been retweeted 86 times.
RT @cleoabram: for my latest video on dog anti-aging (!), I asked the Huge If True audience to send me photos of their dogs as puppies v. o…



The tweet was tweeted at 2022-05-18 12:13:27+00:00.
The original tweet has been retweeted 2902 times.
RT @spiantado: Here is why IQ is bullshit.    A thread.



The tweet was tweeted at 2022-05-18 12:04:11+00:00.
The original tweet has been retweeted 1 times.
Great work, Luke. Thanks to @UMontanaBiz for the support to work on this. https://t.co/29sDihvt2F



The tweet was tweeted at 2022-05-17 12:43:53+00:00.
The original tweet has been retweeted 36 times.
RT @EffinBirds: I was walking @TheWebbyAwards red carpet with Effin’ Birds newsletter writer @michellemanafy when @BigBird showed up.   htt…



The tweet was tweeted at 2022-05-13 23:46:46+00:00.
The original tweet has been retweeted 5 times.
RT @johnhollinger: @NateSilver538 Atlanta still lead

## Pulling Follower Information

In this next section of the assignment, we will pull information about the followers of your two artists. We must first get the follower IDs, then we will be able to "hydrate" the IDs, pulling the user objects for them. Once we have those user objects we will extract some fields that we can use in future analyses. 


The Twitter API only allows users to make 15 requests per 15 minutes when pulling followers. Each request allows you to gather 5000 follower ids. Tweepy will grab the 15 requests quickly then wait 15 minutes, rather than slowly pull the requests over the time period. Before we start grabbing follower IDs, let's first just check how long it would take to pull all of the followers. To do this we use the `followers_count` item from the user object. 

In [27]:
# I'm putting the handles in a list to iterate through below
handles = ['pink', 'dual lipa']

# This will iterate through each Twitter handle that we're collecting from
for screen_name in handles:
    
    # Tells Tweepy we want information on the handle we're collecting from
    # The next line specifies which information we want, which in this case is the number of followers 
    user = api.get_user(screen_name = 'handles') 
    followers_count = user.followers_count

    # Let's see roughly how long it will take to grab all the follower IDs. 
    print(f'''
    @{screen_name} has {followers_count} followers. 
    That will take roughly {followers_count/(5000*15*4):.2f} hours to pull the followers.
    ''')
    


    @pink has 3313 followers. 
    That will take roughly 0.01 hours to pull the followers.
    

    @dual lipa has 3313 followers. 
    That will take roughly 0.01 hours to pull the followers.
    


As we pull data for each artist we will write their data to a folder called "twitter", so we will make that folder if needed.

In [28]:
# Make the "twitter" folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then "unlink" it. Then create a new one.

if not os.path.isdir("twitter") : 
    #shutil.rmtree("twitter/")
    os.mkdir("twitter")

In this following cells, use the `api.followers_ids` (and the `tweepy.Cursor` functionality) to pull some of the followers for your two artists. As you pull the data, write the follower ids to a file called `[artist name]_followers.txt` in the "twitter" folder. For instance, for Cher I would create a file named `cher_followers.txt`. As you pull the data, also store it in an object like a list or a data frame.

In [29]:
num_followers_to_pull = 60*1000 # feel free to use this to limit the number of followers you pull.

In [67]:
# Modify the below code stub to pull the follower IDs and write them to a file. 

# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()

for handle in handles :
    
        output_file = handle + "_followers.txt"
    
    
    # Pull and store the follower IDs
    
followers_ids = []
for page in tweepy.Cursor(api.get_follower_ids, screen_name="handle").pages():
    followers_ids.extend(page)
    time.sleep(60)
        

list = len(followers_ids)
print(list)
    
    # Write the IDs to the output file in the `twitter` folder.
    
with open('twitter.txt','w') as file:
    file.write(str(followers_ids))
        
        
    # If you've pulled num_followers_to_pull, feel free to break out paged twitter API response
            
        
        
# Let's see how long it took to grab all follower IDs
end_time = datetime.datetime.now()
print(end_time - start_time)



0
0:00:00.220779


Now that you have your follower ids, gather some information that we can use in future assignments on them. Using the `lookup_users` function, pull the user objects for your followers. These requests are limited to 900 per 15 minutes, but you can request 100 users at a time. At 90,000 users per 15 minutes, the rate limiter on pulls might be bandwidth rather than API limits. 

Extract the following fields from the user object: 

* screen_name	
* name	
* id	
* location	
* followers_count	
* friends_count	
* description

These can all be accessed via these names in the object. Store the fields with one user per row in a tab-delimited text file with the name `[artist name]_follower_data.txt`. For instance, for Cher I would create a file named `cher_follower_data.txt`. 


In [61]:
# in this cell, do the following
# 1. Set up a data frame or dictionary to hold the user information
# 2. Use the `lookup_users` api function to pull sets of 100 users at a time

# getting the users by user_ids
users = api.lookup_users(screen_name = handle)
  
# printing the user details
for user in users:
    print("The id is : " + str(user.id))
    print("The screen name is : " + user.screen_name, end = "\n\n")

# 3. Store the listed fields in your data frame or dictionary.
# 4. Write the user information in tab-delimited form to the follower data text file. 


The id is : 22005817
The screen name is : u

The id is : 940631
The screen name is : a

The id is : 50465434
The screen name is : L

The id is : 11046
The screen name is : p



One note: the user's description can have tabs or returns in it, so make sure to clean those out of the description before writing them to the file. Here's an example of how you might do this. 

In [31]:
tricky_description = """
    Home by Warsan Shire
    
    no one leaves home unless
    home is the mouth of a shark.
    you only run for the border
    when you see the whole city
    running as well.

"""
# This won't work in a tab-delimited text file.

clean_description = re.sub(r"\s+"," ",tricky_description)
clean_description

' Home by Warsan Shire no one leaves home unless home is the mouth of a shark. you only run for the border when you see the whole city running as well. '

---

# Lyrics Scrape

This section asks you to pull data from the Twitter API and scrape www.AZLyrics.com. In the notebooks where you do that work you are asked to store the data in specific ways. 

In [78]:
# we'll use this dictionary to hold both the artist name and the link on AZlyrics
# my artists are "Pink" & Dua Lipa

artists = {'pink':"https://www.azlyrics.com/p/pink.html",
          'dualipa':"https://www.azlyrics.com/d/dualipa.html"} 

## A Note on Rate Limiting

The lyrics site, www.azlyrics.com, does not have an explicit maximum on number of requests in any one time, but in our testing it appears that too many requests in too short a time will cause the site to stop returning lyrics pages. (Entertainingly, the page that gets returned seems to only have the song title to [a Tom Jones song](https://www.azlyrics.com/lyrics/tomjones/itsnotunusual.html).) 

Whenever you call `requests.get` to retrieve a page, put a `time.sleep(5 + 10*random.random())` on the next line. This will help you not to get blocked. If you _do_ get blocked, which you can identify if the returned pages are not correct, just request a lyrics page through your browser. You'll be asked to perform a CAPTCHA and then your requests should start working again. 

## Part 1: Finding Links to Songs Lyrics

That general artist page has a list of all songs for that artist with links to the individual song pages. 

Q: Take a look at the `robots.txt` page on www.azlyrics.com. (You can read more about these pages [here](https://developers.google.com/search/docs/advanced/robots/intro).) Is the scraping we are about to do allowed or disallowed by this page? How do you know? 

A: Allowed, because our URLs aren't linked from other places on the web.


In [108]:
# Let's set up a dictionary of lists to hold our links
lyrics_pages = defaultdict (lambda:(list))

for artist, artist_page in artists.items() :
    # request the page and sleep
    r = requests.get(artist_page)
    time.sleep(5 + 10*random.random())
    
    # now extract the links to lyrics pages from this page
    # store the links `lyrics_pages` where the key is the artist and the value is a list of links.

url = 'https://www.azlyrics.com/p/pink.html'

links = []

urls = soup.find_all(class_ = 'chartlist-name')

for url in urls:
    relative_link = url.find('a')['href']
    link = 'https://www.azlyrics.com/' + relative_link
    links.append(link)
print(links)
     

SSLError: HTTPSConnectionPool(host='www.azlyrics.com', port=443): Max retries exceeded with url: /p/pink.html (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1076)')))

In [52]:
lyrics_pages

defaultdict(list, {})

Let's make sure we have enough lyrics pages to scrape. 

In [50]:
for artist, lp in lyrics_pages.items() :
    assert(len(set(lp)) > 20) 

In [51]:
# Let's see how long it's going to take to pull these lyrics 
# if we're waiting `5 + 10*random.random()` seconds 
for artist, links in lyrics_pages.items() : 
    print(f"For {artist} we have {len(links)}.")
    print(f"The full pull will take for this artist will take {round(len(links)*10/3600,2)} hours.")

## Part 2: Pulling Lyrics

Now that we have the links to our lyrics pages, let's go scrape them! Here are the steps for this part. 

1. Create an empty folder in our repo called "lyrics". 
1. Iterate over the artists in `lyrics_pages`. 
1. Create a subfolder in lyrics with the artist's name. For instance, if the artist was Cher you'd have `lyrics/cher/` in your repo.
1. Iterate over the pages. 
1. Request the page and extract the lyrics from the returned HTML file using BeautifulSoup.
1. Use the function below, `generate_filename_from_url`, to create a filename based on the lyrics page, then write the lyrics to a text file with that name. 


In [37]:
def generate_filename_from_link(link) :
    
    if not link :
        return None
    
    # drop the http or https and the html
    name = link.replace("https","").replace("http","")
    name = link.replace(".html","")

    name = name.replace("/lyrics/","")
    
    # Replace useless chareacters with UNDERSCORE
    name = name.replace("://","").replace(".","_").replace("/","_")
    
    # tack on .txt
    name = name + ".txt"
    
    return(name)


In [42]:
# Make the lyrics folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then use shutil.rmtree to remove it and create a new one.

import shutil

if os.path.isdir("lyrics") : 
    shutil.rmtree("lyrics/")

os.mkdir("lyrics")

In [88]:
url_stub = "https://www.azlyrics.com" 
start = time.time()

total_pages = 0 

#for artist in lyrics_pages :

    # Use this space to carry out the following steps: 
    
    # 1. Build a subfolder for the artist
    # 2. Iterate over the lyrics pages
    # 3. Request the lyrics page. 
        # Don't forget to add a line like `time.sleep(5 + 10*random.random())`
        # to sleep after making the request
    # 4. Extract the title and lyrics from the page.
    # 5. Write out the title, two returns ('\n'), and the lyrics. Use `generate_filename_from_url`
    #    to generate the filename. 
    
    
    # Remember to pull at least 20 songs per artist. It may be fun to pull all the songs for the artist.

In [ ]:
# extracting the title from the page

r = requests.get(artist_page)
time.sleep(5 + 10*random.random())

soup = BeautifulSoup(r.text, 'html.parser')

for title in soup.find_all('title'):
    title_text = title.text
    title_text = title.text.split('-')[0]
    title_text = title.text.split('|')[0]
    title_text = title.text.replace(' Lyrics', '').strip
    
title_text

In [ ]:
# making the request:
r = requests.get(artist_page)
time.sleep(5 + 10*random.random())

# passing to the soup
soup = BeautifulSoup(r.text, 'html.parser')

soup_boiler = 'Usage of azlyrics.com content by any third-party'

for item in soup.find_all('div'):
    if title_text in item.text:
        break
        print(item.text)

In [ ]:
# extracting the lyrics from the page

lyrics = ''
good_line = False

for line in item.text.split('\n'):
    if title_text in line and not 'lyrics' in line:
        good_line = True
        
    if 'Submit Correction' in line:
        break
        
    if good_line:
        lyrics = lyrics + '\n' + line
        
print(lyrics.strip())

In [ ]:
print(f"Total run time was {round((time.time() - start)/3600,2)} hours.")

---

# Evaluation

This assignment asks you to pull data from the Twitter API and scrape www.AZLyrics.com.  After you have finished the above sections , run all the cells in this notebook. Print this to PDF and submit it, per the instructions.

In [62]:
# Simple word extractor from Peter Norvig: https://norvig.com/spell-correct.html
def words(text): 
    return re.findall(r'\w+', text.lower())

---

## Checking Twitter Data

The output from your Twitter API pull should be two files per artist, stored in files with formats like `cher_followers.txt` (a list of all follower IDs you pulled) and `cher_followers_data.txt`. These files should be in a folder named `twitter` within the repository directory. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [ ]:
twitter_files = os.listdir("twitter")
twitter_files = [f for f in twitter_files if f != ".DS_Store"]
artist_handles = list(set([name.split("_")[0] for name in twitter_files]))

print(f"We see two artist handles: {artist_handles[0]} and {artist_handles[1]}.")

In [ ]:
for artist in artist_handles :
    follower_file = artist + "_followers.txt"
    follower_data_file = artist + "_followers_data.txt"
    
    ids = open("twitter/" + follower_file,'r').readlines()
    
    print(f"We see {len(ids)-1} in your follower file for {artist}, assuming a header row.")
    
    with open("twitter/" + follower_data_file,'r') as infile :
        
        # check the headers
        headers = infile.readline().split("\t")
        
        print(f"In the follower data file ({follower_data_file}) for {artist}, we have these columns:")
        print(" : ".join(headers))
        
        description_words = []
        locations = set()
        
        
        for idx, line in enumerate(infile.readlines()) :
            line = line.strip("\n").split("\t")
            
            try : 
                locations.add(line[3])            
                description_words.extend(words(line[6]))
            except :
                pass
    
        

        print(f"We have {idx+1} data rows for {artist} in the follower data file.")

        print(f"For {artist} we have {len(locations)} unique locations.")

        print(f"For {artist} we have {len(description_words)} words in the descriptions.")
        print("Here are the five most common words:")
        print(Counter(description_words).most_common(5))

        
        print("")
        print("-"*40)
        print("")
    

## Checking Lyrics 

The output from your lyrics scrape should be stored in files located in this path from the directory:
`/lyrics/[Artist Name]/[filename from URL]`. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [72]:
artist_folders = os.listdir("lyrics/")
artist_folders = [f for f in artist_folders if os.path.isdir("lyrics/" + f)]

for artist in artist_folders : 
    artist_files = os.listdir("lyrics/" + artist)
    artist_files = [f for f in artist_files if 'txt' in f or 'csv' in f or 'tsv' in f]

    print(f"For {artist} we have {len(artist_files)} files.")

    artist_words = []

    for f_name in artist_files : 
        with open("lyrics/" + artist + "/" + f_name) as infile : 
            artist_words.extend(words(infile.read()))

            
    print(f"For {artist} we have roughly {len(artist_words)} words, {len(set(artist_words))} are unique.")
